### PR TITLE
Create ExCustomer entity for terminated customer for regulatory requirements

### PR DESCRIPTION
**Pull request by vihangpatil**

### CODE REVIEW

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 1]
```diff
@@ -404,19 +438,55 @@ object Neo4jStoreSingleton : GraphStore {
...
+                val regions = get(Region linkedToCustomer (Customer withId customerId)).bind()
+                // TODO vihang: clear eKYC data for Regions without any SimProfile
+//                val regionsWithoutSimProfile = regions - simProfileRegions
+//                // Link regions with SIM profiles to ExCustomer
+//                for (region in simProfileRegions) {
+//                    fact { (ExCustomer withId customerId) belongedTo (Region withCode region.id) }
+//                }
+                // (For now) Link regions to ExCustomer
+                for (region in regions) {
+                    fact { (ExCustomer withId customerId) belongedTo (Region withCode region.id) }
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Found a missing `bind()`.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 2]
```diff
@@ -404,19 +438,55 @@ object Neo4jStoreSingleton : GraphStore {
...
-                    test = statementResult.summary().counters().nodesDeleted() > 0,
-                    ifTrue = {},
-                    ifFalse = { NotFoundError(type = identityEntity.name, id = identity.id) })
-        }
+        IO {
+            Either.monad<StoreError>().binding {
+                // get customer id
+                val customerId = getCustomerId(identity).bind()
+                // create ex-customer with same id
+                create { ExCustomer(id = customerId, terminationDate = LocalDate.now().toString()) }.bind()
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:447]: Why don't we use timestamp

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 3]
```diff
@@ -404,19 +438,55 @@ object Neo4jStoreSingleton : GraphStore {
...
+                // get all subscriptions and link them to ex-customer
+                val subscriptions = get(Subscription subscribedBy (Customer withId customerId)).bind()
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:449]: `Customer withId customerId` is used in more than one location, move this to a variable

File: neo4j-store/src/main/kotlin/org/ostelco/prime/dsl/Syntax.kt [hunk 4]
```diff
@@ -122,6 +127,24 @@ data class CustomerContext(override val id: String) : EntityContext<Customer>(Cu
...
 }
 
+data class ExCustomerContext(override val id: String) : EntityContext<ExCustomer>(ExCustomer::class, id) {
+
+    infix fun had(simProfile: SimProfileContext) = RelationExpression(
+            relationType = exCustomerToSimProfileRelation,
+            fromId = id,
+            toId = simProfile.id)
+
+    infix fun subscribedTo(subscription: SubscriptionContext) = RelationExpression(
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/dsl/Syntax.kt:137]: Can't we name this `hadSubscribedTo` or `wasSubscribedTo` ??

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 5]
```diff
@@ -404,19 +438,55 @@ object Neo4jStoreSingleton : GraphStore {
...
+                for (subscription in subscriptions) {
+                    fact { (ExCustomer withId customerId) subscribedTo (Subscription withMsisdn subscription.msisdn) }.bind()
+                }
+                // get all SIM profiles and link them to ex-customer.
+                val simProfiles = get(SimProfile forCustomer (Customer withId customerId)).bind()
+                val simProfileRegions = mutableSetOf<Region>()
+                for (simProfile in simProfiles) {
+                    fact { (ExCustomer withId customerId) had (SimProfile withId simProfile.id) }.bind()
+                    // also get regions linked to those SimProfiles.
+                    simProfileRegions.addAll(get(Region linkedToSimProfile (SimProfile withId simProfile.id)).bind())
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:459]: Not used, comment this as well

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 6]
```diff
@@ -404,19 +438,55 @@ object Neo4jStoreSingleton : GraphStore {
...
+//                // Link regions with SIM profiles to ExCustomer
+//                for (region in simProfileRegions) {
+//                    fact { (ExCustomer withId customerId) belongedTo (Region withCode region.id) }.bind()
+//                }
+                // (For now) Link regions to ExCustomer
+                for (region in regions) {
+                    fact { (ExCustomer withId customerId) belongedTo (Region withCode region.id) }.bind()
+                }
+
+                // TODO vihang: When we read and then delete, it fails when deserialization does not work.
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:474]: Purchases, EKYC scans and Identities are removed without backing up. Are we sure ??

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/model/Model.kt [hunk 7]
```diff
@@ -38,3 +38,10 @@ data class Segment(override val id: String) : HasId {
 }
 
 data class Offer(override val id: String) : HasId
+
+data class ExCustomer(
+        override val id:String,
+        val terminationDate: String) : HasId {
```
Comment (prasanthu) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/model/Model.kt:44]: I would use a timestamp

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 2 above)
Comment (vihangpatil) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:447]: As of now, this is a manual process. So, did not want to have additional step of converting epoch long to date.  Also, we have no use of time part.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 3 above)
Comment (vihangpatil) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:449]: Yes, `Customer withId customerId` is a CustomerContext object.
Kept it inline for readability.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/dsl/Syntax.kt (hunk 4 above)
Comment (vihangpatil) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/dsl/Syntax.kt:137]: Yes, that's a better name. I agree.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 6 above)
Comment (vihangpatil) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:474]: We may add Identities and SHA256 of `contactEmail`. Rest of the info is not needed as per requirements.  Also, we might have an option (in future) such as "scheduled for deletion" along with "immediate delete" to address those concerns.  But once deleted, we do not need to store those details unless asked for in requirements.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/model/Model.kt (hunk 7 above)
Comment (vihangpatil) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/model/Model.kt:44]: As mentioned in the comment before, but do not have a strong opinion on it.
```
As of now, this is a manual process. So, did not want to have an additional step of converting epoch long to date. Also, we have no use of time part.
```

**Pull request merged**